sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,539 @@
+  }
+
+  return kPmpRegionConfigureOk;
+}
+
+pmp_region_configure_na4_result_t pmp_region_configure_na4(
+    pmp_region_index_t region, const pmp_region_config_t *config,
+    pmp_region_address_t address) {
+  if (PMP_GRANULARITY_IBEX > 0) {
+    return kPmpRegionConfigureNa4Unavailable;
+  }
+
+  if (config == NULL) {
+    return kPmpRegionConfigureNa4BadArg;
+  }
+
+  if (region >= PMP_REGIONS_NUM) {
+    return kPmpRegionConfigureNa4BadRegion;
+  }
+
+  if (!pmp_address_aligned(address)) {
+    return kPmpRegionConfigureNa4BadAddress;
+  }
+
+  uint32_t field_value = 0;
+  if (!pmp_cfg_permissions_set(config->permissions, &field_value)) {
+    return kPmpRegionConfigureNa4Error;
+  }
+
+  pmp_cfg_mode_lock_set(config->lock, &field_value);
+
+  field_value = bitfield_field32_write(field_value, kPmpCfgModeField,
+                                       PMP_CFG_CSR_MODE_NA4);
+
+  // Address registers must be written prior to the configuration registers to
+  // ensure that they are not locked.
+  pmp_region_configure_result_t result = pmp_csr_address_write(region, address);
+  if (result != kPmpRegionConfigureOk) {
+    return (pmp_region_configure_na4_result_t)result;
+  }
+
+  result = pmp_csr_cfg_field_write(region, field_value);
+  if (result != kPmpRegionConfigureOk) {
+    return (pmp_region_configure_na4_result_t)result;
+  }
+
+  return kPmpRegionConfigureNa4Ok;
+}
+
+pmp_region_configure_napot_result_t pmp_region_configure_napot(
+    pmp_region_index_t region, const pmp_region_config_t *config,
+    pmp_region_address_t address, uint32_t size) {
+  if (config == NULL) {
+    return kPmpRegionConfigureNapotBadArg;
+  }
+
+  if (region >= PMP_REGIONS_NUM) {
+    return kPmpRegionConfigureNapotBadRegion;
+  }
+
+  pmp_region_address_t napot_address;
+  pmp_region_configure_napot_result_t napot_result =
+      pmp_napot_address_construct(address, size, &napot_address);
+  if (napot_result != kPmpRegionConfigureNapotOk) {
+    return napot_result;
+  }
+
+  uint32_t field_value = 0;
+  if (!pmp_cfg_permissions_set(config->permissions, &field_value)) {
+    return kPmpRegionConfigureNapotError;
+  }
+
+  pmp_cfg_mode_lock_set(config->lock, &field_value);
+
+  field_value = bitfield_field32_write(field_value, kPmpCfgModeField,
+                                       PMP_CFG_CSR_MODE_NAPOT);
+
+  // Address registers must be written prior to the configuration registers to
+  // ensure that they are not locked.
+  pmp_region_configure_result_t result =
+      pmp_csr_address_write(region, napot_address);
+  if (result != kPmpRegionConfigureOk) {
+    return (pmp_region_configure_napot_result_t)result;
+  }
+
+  result = pmp_csr_cfg_field_write(region, field_value);
+  if (result != kPmpRegionConfigureOk) {
+    return (pmp_region_configure_napot_result_t)result;
+  }
+
+  return kPmpRegionConfigureNapotOk;
+}
+
+pmp_region_configure_result_t pmp_region_configure_tor(
+    pmp_region_index_t region, const pmp_region_config_t *config,
```

Comment:
It may be helpful to call this sthg like `end_region`.

Reply:
Makes sense, thanks :)

Reply:
Done.